### ISSUE TITLE
Rule proposal: `no-useless-undefined`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#683]
# Fail

```js
function foo() {
	return undefind;
}
```

```js
let a = undefined;
```

```js
call(undefined);
```

```js
call(foo, bar, undefined);
```

```js
const {foo = undefined} = {};
```

```js
function foo({bar = undefined} = {}) {}
```

```js
function foo(bar = undefined) {}
```

# Pass

```js
function foo() {
	return;
}
```

```js
let a;
```

```js
call();
```

```js
call(foo, bar);
call(foo, bar,);
```

```js
const {foo} = {};
```

```js
function foo({bar} = {}) {}
```

```js
function foo(bar) {}
```

Discussion https://github.com/sindresorhus/eslint-plugin-unicorn/pull/636#discussion_r407150865

Not sure about this, 

```js
const noop = () => undefined;

// prefer this?
const noop = () => {};
```

**Comment by sindresorhus** [comment]
> // prefer this?
> const noop = () => {};

Yes

**Comment by sindresorhus** [comment]
This is accepted.

**Comment by snowteamer** [comment]
I think this rule is a great idea, but how are we supposed to mark a formal parameter as optional if the following form is disallowed ?
```js
function foo(bar = undefined) {
  if (typeof bar !== "undefined") {
    // Do something with 'bar'
  }
}
```
- Omitting the parameter declaration altogether does not seem to be an option since the function body will probably use it ;
- It is unclear to me wheather replacing `= undefined` with `= void 0` is a good thing.

This information can of course be expressed in annotations, but I would prefer to be able to also keep it in the code.

**Comment by fisker** [comment]
We  are not removing parameter `bar`

```js
function foo(bar = undefined) {
  if (typeof bar !== "undefined") {
    // Do something with 'bar'
  }
}
```

To

```js
function foo(bar) {
  if (typeof bar !== "undefined") {
    // Do something with 'bar'
  }
}
```

Is this breaking something?

**Comment by snowteamer** [comment]
It does not break anything *per se*,  but it still remove the visual indication that the `bar` parameter is optional, which is the reason why I sometimes use that form.

**Comment by fisker** [comment]
How about

```js
function foo(bar /* optional */ ) {}
```

**Comment by snowteamer** [comment]
I think this is an interesting solution, I might give it a try if nothing else changes - thank you for your time.

**Comment by fisker** [comment]
Not sure which one better, but all  seems  clear
```js
function foo(bar, baz /* optional */) {}
```
```js
function foo(bar, /* optional */ baz) {}
```

**Comment by fisker** [comment]
reopen by  accident.
